src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,1084 @@
+
+            public OrCases()
+            {
+                Cases = ArrayBuilder<(BoundPattern pattern, SyntaxNode? syntax)>.GetInstance();
+            }
+
+            public void Add(BoundPattern pattern, SyntaxNode? syntax)
+            {
+                Cases.Add((pattern, pattern.WasCompilerGenerated ? null : syntax));
+            }
+
+            public void Dispose()
+            {
+                Cases.Free();
+            }
+        }
+
+        // The purpose of this rewriter is to push `not` patterns down the pattern tree and
+        // pull all the `and` and `or` patterns up the pattern tree.
+        // It needs to expand composite patterns in the process.
+        // It also erases/simplifies some patterns (variable declarations).
+        // Throughout the process it maintains input and narrowed types discipline, to produce a valid and consistent output.
+        private class PatternNormalizer : BoundTreeRewriterWithStackGuard
+        {
+            private bool _negated;
+            private TypeSymbol _inputType;
+            private readonly Conversions _conversions;
+
+            public PatternNormalizer(Conversions conversions, TypeSymbol inputType)
+            {
+                _negated = false;
+                _inputType = inputType;
+                _conversions = conversions;
+            }
+
+            internal static BoundPattern Rewrite(BoundPattern pattern, TypeSymbol inputType, Conversions conversions)
+            {
+                return (BoundPattern)new PatternNormalizer(conversions, inputType).Visit(pattern);
+            }
+
+            public override BoundNode? VisitBinaryPattern(BoundBinaryPattern node)
+            {
+                var resultLeft = VisitWithInputType(node.Left, _inputType);
+                var resultDisjunction = _negated ? !node.Disjunction : node.Disjunction;
+
+                var rightInputType = resultDisjunction ? _inputType : resultLeft.NarrowedType;
+                var resultRight = VisitWithInputType(node.Right, rightInputType);
+
+                TypeSymbol narrowedTypeForBinary2 = NarrowedTypeForBinary(resultLeft, resultRight, resultDisjunction);
+                var result = new BoundBinaryPattern(node.Syntax, resultDisjunction, resultLeft, resultRight, _inputType, narrowedTypeForBinary2);
+                return result;
+            }
+
+            private TypeSymbol NarrowedTypeForBinary(BoundPattern resultLeft, BoundPattern resultRight, bool resultDisjunction)
+            {
+                TypeSymbol narrowedType;
+                if (resultDisjunction)
+                {
+                    var candidateTypes = ArrayBuilder<TypeSymbol>.GetInstance(2);
+                    candidateTypes.Clear();
+                    candidateTypes.Add(resultLeft.NarrowedType);
+                    candidateTypes.Add(resultRight.NarrowedType);
+
+                    narrowedType = this.LeastSpecificType(candidateTypes) ?? _inputType;
```

Comment:
>narrowedType = this.LeastSpecificType(candidateTypes) ?? _inputType; [](http://example.com/codeflow?start=20&length=68)

Could we use a simpler logic on this branch? For example:
```
                    if (resultRight.NarrowedType.Equals(resultLeft.NarrowedType, TypeCompareKind.AllIgnoreOptions))
                    {
                        return resultLeft.NarrowedType;
                    }

                    return _inputType;
```

Do we really have to be a 100% accurate with what the initial binding is doing?
Because if we do have to be a 100% accurate, I don't think we are. It looks like
the initial binding calculates a least specific type across all leaves in the entire disjunction
hierarchy, here we are using types from two immediate children. I assume
there is a chance of getting a different result, otherwise why initial binding would do
extra work.